test/schema.service.test.ts

```diff
@@ -0,0 +1,844 @@
+  t.is(associations[0], 0);
+});
+test('reference array properties get', t => {
+  const schema = {
+    definitions: {
+      class: {
+        type: 'object',
+        properties: {
+          id: {
+            type: 'string'
+          },
+          associations: {
+            type: 'array',
+            items: {
+              type: 'integer',
+              minimum: 0
+            }
+          }
+        },
+        links: [{
+          rel: 'full',
+          href: '#/classes/{associations}',
+          targetSchema: '#/definitions/class'
+        }]
+      }
+    },
+    type: 'object',
+    properties: {
+      classes: {
+        type: 'array',
+        items: {
+          $ref: '#/definitions/class'
+        }
+      }
+    }
+  } as JsonSchema;
+  const service: SchemaService = new SchemaServiceImpl(schema);
+  const property =
+    service.getReferenceProperties(schema.definitions.class)[0];
+  const data = {classes: [{id: 1}, {id: 2, associations: [0]}]};
+  const getData = property.getData(data, data.classes[1]);
+  t.is(getData, data.classes[0]);
+});
+
+test('property type check', t => {
+  const schema = {
+    definitions: {
+      class: {
+        type: 'object',
+        properties: {
+          id: {
+            type: 'string'
+          },
+          association: {
+            type: 'integer',
+            minimum: 0
+          }
+        },
+        links: [{
+          rel: 'full',
+          href: '#/classes/{association}',
+          targetSchema: '#/definitions/class'
+        }]
+      }
+    },
+    type: 'object',
+    properties: {
+      classes: {
+        type: 'array',
+        items: {
+          $ref: '#/definitions/class'
+        }
+      }
+    }
+  } as JsonSchema;
+  const service: SchemaService = new SchemaServiceImpl(schema);
+  const refProperty =
+    service.getReferenceProperties(schema.properties.classes.items as JsonSchema)[0];
+  t.true(isRefrenceProperty(refProperty));
+  t.false(isContainmentProperty(refProperty));
+  const containmentProperty = service.getContainmentProperties(schema)[0];
+  t.false(isRefrenceProperty(containmentProperty));
+  t.true(isContainmentProperty(containmentProperty));
+});
+test('self contained child schemata: cross recursion', t => {
+  const schema = {
+    definitions: {
+      person: {
+        type: 'object',
+        properties: {
+          robots: {
+            type: 'array',
+            items: {$ref: '#/definitions/robot'}
+          }
+        }
+      },
+      robot: {
+        type: 'object',
+        properties: {
+          humans: {
+            type: 'array',
+            items: {$ref: '#/definitions/person'}
+          }
+        }
+      }
+    },
+    type: 'object',
+    properties: {
+      persons: {type: 'array', items: {$ref: '#/definitions/person'}}
+    }
+  } as JsonSchema;
+  const service: SchemaService = new SchemaServiceImpl(schema);
+  const properties = service.getContainmentProperties(schema);
```

Comment:
Two things I noted when calling this method:
* The method will modify the original schema that has been passed in. This should be mentioned in the JSDoc
* The generated ID field in this contains two hashes and looks like this `##/definitions/person`. I don't think that's on purpose